File: csharp/interest-is-interesting/InterestIsInteresting.cs
using System;

static class SavingsAccount
{
    public static float InterestRate(decimal balance) =>
        balance switch
        {
            < 0 => 3.213f,
            < 1000 => 0.5f,
            >= 5000 => 2.475f,
            >= 1000 => 1.621f
        };

    public static decimal Interest(decimal balance) => balance * (decimal)(InterestRate(balance) / 100);

    public static decimal AnnualBalanceUpdate(decimal balance) => balance + Interest(balance);

    public static int YearsBeforeDesiredBalance(decimal balance, decimal targetBalance)
    {
        var y = 0;
        while (balance < targetBalance)
        {
            balance = AnnualBalanceUpdate(balance);
            ++y;
        }

        return y;
    }
}

File: fsharp/interest-is-interesting/InterestIsInteresting.fs
module InterestIsInteresting

let interestRate (balance: decimal): single =
  if balance >= 5000m then
        2.475f
    elif balance >= 1000m then
        1.621f 
    elif balance >= 0m then
        0.5f
    else
        3.213f

let interest (balance: decimal): decimal =
   let rate = interestRate balance
   balance * decimal rate * 0.01m

let annualBalanceUpdate(balance: decimal): decimal =
   balance + interest balance

let amountToDonate(balance: decimal) (taxFreePercentage: float): int =
   if balance < 0m then
       0
   else
       int(floor (balance * decimal (taxFreePercentage * 0.02)))
